src/main/scala/com/databricks/spark/csv/util/InferSchema.scala

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2014 Databricks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.databricks.spark.csv.util
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.analysis.HiveTypeCoercion
+import org.apache.spark.sql.types._
+
+import scala.util.control.Exception._
+
+private[csv] object InferSchema {
```

Comment:
thanks. schema inference would be awesome...i wonder what the performance would be like, as exceptions can be expensive. on the other hand you can see in bigdf code https://github.com/AyasdiOpenSource/bigdf/blob/master/src/main/scala/com/ayasdi/bigdf/SchemaUtils.scala this is what i did :-) an option to use sampling instead of using the whole dataset would be nice. i am happy to extract that code from bigdf and contribute

possible optimizations:
1) let inferField know what we have so far, so it can avoid doing unnecessary work. e.g. if we got a double, no need to try parsing as int or long. if we got a string, no need to parse as double.
2) use mappartitions to do an "imperative-style" loop per partition and then merge the partitions (which can be done on the driver as there will be a single schema per partition). this will reduce memory allocation and GC overhead.